File: sdks/typescript/src/trigger.ts
// Producer-side trigger client — an idiomatic wrapper over `POST /v1/trigger`.
//
// Parity with the .NET `ICroniqTriggerClient` (#277): fire a registered job on
// demand so the *same* handler serves both its Croniqfile schedule and
// near-real-time, event-driven execution. Deliberately independent of the
// runner: triggering needs the `jobs:trigger` (or `admin`) scope, which runner
// poll keys typically do not carry, so the trigger client takes ITS OWN
// credentials rather than reusing a runner's.

import { HttpError } from './client.js';
import { AbortError } from './deferred.js';
import { composeSignals, isAbortLikeError } from './http.js';
import type { TriggerRequest, TriggerResponse } from './protocol.js';
import { trimTrailingSlashes } from './url.js';

/** Options for {@link CroniqTriggerClient} / {@link createTriggerClient}. */
export interface CroniqTriggerClientOptions {
  /** Base URL of the Croniq server, e.g. `http://localhost:4000`. */
  serverUrl: string;

  /**
   * API key for `Authorization: ApiKey {key}`. Takes precedence over
   * {@link bearerToken} when both are set. Needs the `jobs:trigger` (or
   * `admin`) scope — distinct from a runner's poll scopes.
   */
  apiKey?: string | undefined;

  /** Bearer token for `Authorization: Bearer {token}`. */
  bearerToken?: string | undefined;

  /** Per-request timeout in milliseconds. Default 30 000 ms. */
  requestTimeoutMs?: number | undefined;

  /** Custom `fetch` implementation. Defaults to the global `fetch`. */
  fetchImpl?: typeof fetch | undefined;
}

/** Optional arguments for a single {@link CroniqTriggerClient.trigger} call. */
export interface TriggerParams {
  /**
   * Arbitrary JSON metadata forwarded to the handler as-is (merged over the
   * job's DSL metadata server-side). Keys starting with `__` are reserved.
   */
  metadata?: Record<string, unknown>;

  /** Capabilities a runner MUST have to be assigned this execution. */
  require?: string[];

  /** Capabilities used to prefer runners when several are eligible. */
  prefer?: string[];

  /**
   * Execution timeout as a server duration string (e.g. `"30s"`, `"5m"`).
   * The server default applies when omitted.
   */
  timeout?: string;

  /**
   * Optional dedup key (≤ 200 chars, scoped per `job_key`). Servers with
   * trigger-idempotency support (#279) coalesce repeat triggers carrying the
   * same key onto the existing execution (see {@link TriggerResult.deduplicated});
   * older servers ignore it.
   */
  idempotencyKey?: string;

  /** Abort signal to cancel the underlying HTTP call. */
  signal?: AbortSignal;
}

/** Result of an on-demand job trigger (`POST /v1/trigger`). */
export interface TriggerResult {
  /**
   * Identifier of the execution the trigger resolved to. On a dedup hit this
   * is the EXISTING execution's id, not a new one.
   */
  executionId: string;

  /** Server work-queue depth after the trigger was processed. */
  queued: number;

  /**
   * `true` when the server coalesced this trigger onto an existing execution
   * because the request carried an `idempotency_key` it had already seen.
   * Always `false` on servers without idempotency-key support (they omit the
   * flag and the client defaults it to `false`).
   */
  deduplicated: boolean;
}

/**
 * Thrown when `POST /v1/trigger` is rejected with `429 Too Many Requests`
 * because the job's per-job queue-depth cap is reached (#299). A producer that
 * batches or retries triggers should treat this as backpressure — back off and
 * retry later rather than dropping the work silently. {@link retryAfterMs}
 * carries the server's `Retry-After` hint when present.
 */
export class QueueOverflowError extends HttpError {
  constructor(
    path: string,
    body: string,
    public readonly retryAfterMs?: number,
  ) {
    super(429, 'Too Many Requests', path, body);
    this.name = 'QueueOverflowError';
  }
}

const TRIGGER_PATH = '/v1/trigger';
const DEFAULT_REQUEST_TIMEOUT_MS = 30_000;

/**
 * Client for firing Croniq jobs on demand. Serialises a snake_case JSON body,
 * attaches `Authorization` per request (ApiKey precedence over Bearer), and
 * parses the `TriggerResponse` (defaulting a missing `deduplicated` to
 * `false`). Non-2xx responses surface as errors — `429` as a
 * {@link QueueOverflowError}, everything else as an {@link HttpError}.
 */
export class CroniqTriggerClient {
  readonly #baseUrl: string;
  readonly #apiKey: string | undefined;
  readonly #bearerToken: string | undefined;
  readonly #requestTimeoutMs: number;
  readonly #fetch: typeof fetch;

  constructor(opts: CroniqTriggerClientOptions) {
    if (!opts.serverUrl || typeof opts.serverUrl !== 'string') {
      throw new TypeError('CroniqTriggerClientOptions.serverUrl is required');
    }
    try {
      // eslint-disable-next-line no-new
      new URL(opts.serverUrl);
    } catch {
      throw new TypeError(`CroniqTriggerClientOptions.serverUrl is not a valid URL: ${opts.serverUrl}`);
    }
    this.#baseUrl = trimTrailingSlashes(opts.serverUrl);
    this.#apiKey = opts.apiKey;
    this.#bearerToken = opts.bearerToken;
    this.#requestTimeoutMs = opts.requestTimeoutMs ?? DEFAULT_REQUEST_TIMEOUT_MS;
    this.#fetch = opts.fetchImpl ?? fetch;
  }

  /**
   * Fire a job immediately. Its registered handler runs on the next eligible
   * runner, exactly like a scheduled fire.
   *
   * @param jobKey Job key, e.g. `billing:invoice-generate`.
   * @param params Optional metadata / routing hints / timeout / idempotency key.
   * @returns The created (or deduplicated) execution and queue depth.
   * @throws {QueueOverflowError} on `429` per-job queue overflow (#299).
   * @throws {HttpError} on any other non-2xx response.
   */
  async trigger(jobKey: string, params: TriggerParams = {}): Promise<TriggerResult> {
    if (typeof jobKey !== 'string' || jobKey.trim().length === 0) {
      throw new TypeError('trigger(jobKey): jobKey must be a non-empty string');
    }

    // Only assign supplied optionals: JSON.stringify drops `undefined`, so an
    // unset field never reaches the wire.
    const requestBody: TriggerRequest = { job_key: jobKey };
    if (params.metadata !== undefined) requestBody.metadata = params.metadata;
    if (params.require !== undefined) requestBody.require = params.require;
    if (params.prefer !== undefined) requestBody.prefer = params.prefer;
    if (params.timeout !== undefined) requestBody.timeout = params.timeout;
    if (params.idempotencyKey !== undefined) requestBody.idempotency_key = params.idempotencyKey;

    const headers: Record<string, string> = {
      'content-type': 'application/json',
      accept: 'application/json',
    };
    if (this.#apiKey) {
      headers.authorization = `ApiKey ${this.#apiKey}`;
    } else if (this.#bearerToken) {
      headers.authorization = `Bearer ${this.#bearerToken}`;
    }

    const { signal, dispose } = composeSignals(params.signal, this.#requestTimeoutMs);

    let response: Response;
    try {
      response = await this.#fetch(`${this.#baseUrl}${TRIGGER_PATH}`, {
        method: 'POST',
        headers,
        body: JSON.stringify(requestBody),
        signal,
      });
    } catch (err) {
      if (isAbortLikeError(err)) {
        throw new AbortError(`request to ${TRIGGER_PATH} aborted`);
      }
      throw err;
    } finally {
      dispose();
    }

    if (!response.ok) {
      const text = await response.text().catch(() => '');
      if (response.status === 429) {
        throw new QueueOverflowError(TRIGGER_PATH, text, parseRetryAfterMs(response.headers.get('retry-after')));
      }
      throw new HttpError(response.status, response.statusText, TRIGGER_PATH, text);
    }

    const text = await response.text();
    if (text.length === 0) {
      throw new HttpError(response.status, response.statusText, TRIGGER_PATH, 'empty response body');
    }
    let parsed: TriggerResponse;
    try {
      parsed = JSON.parse(text) as TriggerResponse;
    } catch {
      throw new HttpError(response.status, response.statusText, TRIGGER_PATH, `non-JSON response body: ${text}`);
    }

    return {
      executionId: parsed.execution_id,
      queued: parsed.queued,
      deduplicated: parsed.deduplicated ?? false,
    };
  }
}

/** Convenience factory mirroring `createRunner`. */
export function createTriggerClient(opts: CroniqTriggerClientOptions): CroniqTriggerClient {
  return new CroniqTriggerClient(opts);
}

/**
 * Parse a `Retry-After` header value into milliseconds. Supports both the
 * delta-seconds form (`"30"`) and the HTTP-date form. Returns `undefined` when
 * the header is absent or unparseable.
 */
function parseRetryAfterMs(headerValue: string | null): number | undefined {
  if (!headerValue) return undefined;
  const trimmed = headerValue.trim();
  if (trimmed.length === 0) return undefined;
  const seconds = Number(trimmed);
  if (Number.isFinite(seconds) && seconds >= 0) {
    return Math.round(seconds * 1000);
  }
  const dateMs = Date.parse(trimmed);
  if (!Number.isNaN(dateMs)) {
    return Math.max(0, dateMs - Date.now());
  }
  return undefined;
}

File: sdks/go/trigger.go
package croniq

import (
	"context"
	"errors"
	"net/http"
	"strings"
	"time"
)

// DefaultTriggerRequestTimeout bounds a single POST /v1/trigger call when
// the context passed to [TriggerClient.Trigger] carries no deadline of its
// own. Matches the .NET SDK's CroniqClientOptions.RequestTimeout so the two
// producer clients behave the same out of the box.
const DefaultTriggerRequestTimeout = 30 * time.Second

// TriggerRequest is the body sent to POST /v1/trigger.
//
// Optional fields use `omitempty` so a value the caller never supplied is
// omitted from the JSON entirely rather than sent as null/empty: the server
// distinguishes "unset" (inherit the job's configured value) from an explicit
// value, and a producer must not fabricate defaults on the wire.
//
// `omitempty` also drops an EXPLICITLY empty value — a nil-or-empty slice, a
// blank string — and that is the intended contract, not an artefact (issue
// #553). The server already reads an empty Require as "inherit the job's
// runner { require … }", so "require": [] would only be a second wire spelling
// of a message that has one. And Timeout "" is not a parseable duration:
// sending it would hand the runner a broken value where omitting it inherits
// the job's own timeout. The other SDKs normalize empty to absent to match.
type TriggerRequest struct {
	// JobKey is the job to fire, e.g. "billing:invoice". Required.
	JobKey string `json:"job_key"`

	// Metadata is arbitrary caller JSON, merged over the job's DSL metadata
	// and forwarded to the handler verbatim. Nested objects and non-string
	// values are preserved — it is NOT flattened or stringified. Keys
	// starting with "__" are reserved for internal use.
	Metadata map[string]any `json:"metadata,omitempty"`

	// Require lists capabilities a runner MUST have to be assigned this
	// execution. Empty → inherit the job's runner { require … }.
	Require []string `json:"require,omitempty"`

	// Prefer lists capabilities used to prefer runners when several are
	// eligible.
	Prefer []string `json:"prefer,omitempty"`

	// Timeout is the execution timeout as a server duration string
	// (e.g. "30s", "5m"). Empty → inherit the job's configured timeout; the
	// server falls back to 5m only when the job declares none either.
	Timeout string `json:"timeout,omitempty"`

	// IdempotencyKey is an optional dedup key scoped per job_key. A server
	// with trigger-idempotency support coalesces a repeat trigger carrying
	// the same (job_key, idempotency_key) onto the existing execution (see
	// [TriggerResponse.Deduplicated]); older servers ignore it. Capped at
	// 200 characters server-side — a longer key is rejected with 400.
	IdempotencyKey string `json:"idempotency_key,omitempty"`
}

// TriggerResponse is the parsed body of a successful POST /v1/trigger.
type TriggerResponse struct {
	// ExecutionID identifies the execution the trigger resolved to. On a
	// dedup hit this is the EXISTING execution's id, not a new one.
	ExecutionID string `json:"execution_id"`

	// Queued is the server work-queue depth after the trigger was processed
	// (unchanged on a dedup hit — nothing is enqueued then).
	Queued int `json:"queued"`

	// Deduplicated is true when the server coalesced this trigger onto an
	// existing execution because the request carried an idempotency_key it
	// had already seen. Servers without idempotency-key support omit the
	// field entirely; it then decodes to false (Go's zero value), which is
	// the intended default.
	Deduplicated bool `json:"deduplicated"`
}

// TriggerClient is the producer-side client for firing Croniq jobs on
// demand via POST /v1/trigger. It is the counterpart to [Runner] (the
// consumer side) and is deliberately independent of it: triggering requires
// the jobs:trigger (or admin) scope, which runner poll keys typically do
// not carry, so a TriggerClient authenticates with ITS OWN credentials.
//
// Construct one with [NewTriggerClient] and configure auth via the builder
// methods. Safe for concurrent use.
type TriggerClient struct {
	client         *Client
	requestTimeout time.Duration
}

// NewTriggerClient constructs a TriggerClient targeting serverURL (trailing
// slashes tolerated). Each call is bounded by [DefaultTriggerRequestTimeout]
// unless the caller's context already carries a deadline; override with
// [TriggerClient.WithRequestTimeout]. Configure credentials with
// [TriggerClient.WithAPIKey] or [TriggerClient.WithBearer] — the endpoint
// requires the jobs:trigger or admin scope.
//
// serverURL must be https:// unless its host is loopback (localhost,
// 127.0.0.0/8, ::1); a non-loopback http:// URL makes every
// [TriggerClient.Trigger] call fail with a configuration error unless
// [TriggerClient.WithInsecureHTTP] is chained on. See [TriggerClient.Err].
func NewTriggerClient(serverURL string) *TriggerClient {
	return &TriggerClient{
		client:         NewClient(serverURL),
		requestTimeout: DefaultTriggerRequestTimeout,
	}
}

// WithAPIKey configures `Authorization: ApiKey {key}` on every trigger
// request. Returns the same client for chaining.
func (tc *TriggerClient) WithAPIKey(key string) *TriggerClient {
	tc.client.WithAPIKey(key)
	return tc
}

// WithBearer configures `Authorization: Bearer {token}` on every trigger
// request.
func (tc *TriggerClient) WithBearer(token string) *TriggerClient {
	tc.client.WithBearer(token)
	return tc
}

// WithHTTPClient injects a custom *http.Client (for tests, custom
// transports, proxies, mTLS, …).
func (tc *TriggerClient) WithHTTPClient(hc *http.Client) *TriggerClient {
	tc.client.WithHTTPClient(hc)
	return tc
}

// WithInsecureHTTP opts this client in to a cleartext http:// server URL on a
// non-loopback host. Without it such a URL makes every trigger call fail; with
// it the client works but logs one loud warning, because the credential then
// travels in cleartext on every call (and through any HTTP proxy the
// environment configures). Lab and staging only — never production.
func (tc *TriggerClient) WithInsecureHTTP() *TriggerClient {
	tc.client.WithInsecureHTTP()
	return tc
}

// Err reports the base-URL validation failure recorded at construction, or
// nil when the configuration is sound.
func (tc *TriggerClient) Err() error { return tc.client.Err() }

// WithRequestTimeout overrides the per-call timeout applied when the context
// passed to [TriggerClient.Trigger] carries no deadline of its own. A
// non-positive duration is ignored (the default is kept).
func (tc *TriggerClient) WithRequestTimeout(d time.Duration) *TriggerClient {
	if d > 0 {
		tc.requestTimeout = d
	}
	return tc
}

// Trigger fires a job immediately via POST /v1/trigger. The job's registered
// handler runs on the next eligible runner, exactly like a scheduled fire.
//
// Unset optional fields on req are omitted from the request body (never sent
// as null). A non-2xx response — including the 429 the server returns when a
// job is at its per-job queue-overflow cap (max_queue_depth) — is returned
// as a [*ServerError]; a producer batching or retrying triggers should
// inspect [ServerError.Status] to observe backpressure rather than treating
// every failure alike.
//
// When ctx carries no deadline, the call is bounded by the client's request
// timeout (see [TriggerClient.WithRequestTimeout]).
func (tc *TriggerClient) Trigger(ctx context.Context, req *TriggerRequest) (*TriggerResponse, error) {
	if req == nil {
		return nil, errors.New("croniq: trigger request must not be nil")
	}
	if strings.TrimSpace(req.JobKey) == "" {
		return nil, errors.New("croniq: trigger job_key must not be empty")
	}

	if tc.requestTimeout > 0 {
		if _, ok := ctx.Deadline(); !ok {
			var cancel context.CancelFunc
			ctx, cancel = context.WithTimeout(ctx, tc.requestTimeout)
			defer cancel()
		}
	}

	var resp TriggerResponse
	if err := tc.client.do(ctx, http.MethodPost, "/v1/trigger", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}
